src/VisualStudio/Core/Def/Implementation/ChangeSignature/ChangeSignatureDialogViewModel.cs

```diff
@@ -439,25 +527,136 @@ public string RestoreAutomationText
+
+                    text += $"{Modifier} {Type} {Parameter}";
+
+                    if (!string.IsNullOrWhiteSpace(Default))
+                    {
+                        text += $" = {Default}";
+                    }
+
+                    text += string.Format(ServicesVSResources.Inserting_call_site_value_0, Callsite);
+
+                    return text;
+                }
+            }
+
+            public override bool IsDisabled => false;
+            public override string Callsite
+            {
+                get
+                {
+                    if (!string.IsNullOrWhiteSpace(_addedParameter.CallsiteValue))
+                    {
+                        return _addedParameter.CallsiteValue;
+                    }
+
+                    return ServicesVSResources.ChangeSignature_NewParameterIntroduceTODOVariable;
+                }
             }
+
+            internal override Parameter CreateParameter()
+                => new AddedParameter(Type, Parameter, Callsite);
+
+            public override string InitialIndex => ServicesVSResources.ChangeSignature_NewParameterIndicator;
+
+            // Newly added parameters cannot have modifiers yet
+            public override string Modifier => string.Empty;
+
+            // Only required parameters are supported currently
+            public override string Default => string.Empty;
+
+            // Use null as a marker that this is an added parameter
+            public override IParameterSymbol ParameterSymbol => null;
         }
 
-        public class ParameterViewModel
+        public class ExistingParameterViewModel : ParameterViewModel
         {
-            private readonly ChangeSignatureDialogViewModel _changeSignatureDialogViewModel;
+            public override IParameterSymbol ParameterSymbol { get; }
+
+            public ExistingParameterViewModel(ChangeSignatureDialogViewModel changeSignatureDialogViewModel, Parameter parameter, int initialIndex)
+                : base(changeSignatureDialogViewModel)
+            {
+                ParameterSymbol = (parameter as ExistingParameter).Symbol;
+                InitialIndex = initialIndex.ToString();
+            }
 
-            public IParameterSymbol ParameterSymbol { get; }
+            internal override Parameter CreateParameter()
+            {
+                return new ExistingParameter(ParameterSymbol);
+            }
 
-            public ParameterViewModel(ChangeSignatureDialogViewModel changeSignatureDialogViewModel, IParameterSymbol parameter)
+            public override string ShortAutomationText => $"{Type} {Parameter}";
+            public override string FullAutomationText
             {
-                _changeSignatureDialogViewModel = changeSignatureDialogViewModel;
-                ParameterSymbol = parameter;
+                get
+                {
+                    var text = $"{Modifier} {Type} {Parameter}";
+                    if (!string.IsNullOrWhiteSpace(Default))
```

Comment:
Move this to the base and have just the add parameter form call the base and just add the text it wants?